### PR TITLE
Date/time support for Postgres. Year/month/day operations on Columns.

### LINKED ISSUE #5268: Sort out Date and Time literals with time-zone and nanoseconds
This task is automatically imported from the old Task Issue Board and it was originally created by James Dunkerley.
Original issue is [here](https://www.pivotaltracker.com/story/show/183734954).
<hr />

Current implementation does not serialise the fractional part of seconds when create a constant column within the database.

Look at what is possible with JDBC to allow better serialisation rather than fixed format used currently.

### LINKED ISSUE #6115: Add Date/Time support to Postgres backend and implement basic year/month/date operations.
- [x] Extend the `Column_Fetcher` and `Statement_Setter` logic for Postgres dialect to correctly deal with the 3 of our date/time types.
- [x] Add tests with uploading a date/times to a Postgres DB and fetching them back.
	- If needed, also INSERT using a raw query.
- [x] Implement `year`, `month`, `day` operations on `Column`.
	- [x] In in-memory.
	- [x] In the Postgres backend.
	- [x] They should check the input type and issue `Invalid_Value_Type` if it's invalid.
	- They do not accept `Mixed` type columns.
	- ~~They should throw `Unsupported_Database_Operation` in the SQLite backend.~~
		- Not applicable really, since we cannot even construct a Date column in SQLite at the current moment, so we will always just get `Invalid_Value_Type`.
- [x] Add tests for all backends, including checking for Unsupported errors in SQLite.
	- (Use a supported flag and put all tests in `Common_Table_Operations`. If enabled run tests, if disabled - expect errors.
	- [x] Also add tests for expressions using year/month/date operations.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Closes #6115

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Test/0.0.0-dev/src/Extensions.enso [hunk 1]
```diff
@@ -176,6 +176,22 @@ Any.should_start_with self that frames_to_skip=0 = case self of
...
+     displayed as the source of this error.
+
+   > Example
+     Assert that one value should start with another.
+
+         from Standard.Test import Test
+
+         example_should_start_with = "Hello World!" . should_start_with "Hello"
+Error.should_start_with : Any -> Integer -> Test_Result
+Error.should_start_with self _ frames_to_skip=0 = Test.fail_match_on_unexpected_error self 1+frames_to_skip
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Test/0.0.0-dev/src/Extensions.enso:205]: Why does this always fail?

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Test/0.0.0-dev/src/Extensions.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Test/0.0.0-dev/src/Extensions.enso:205]: Because it is called on an error value and a non-error was expected.

This is due to a slightly weird calling convention.

If I call a method on a dataflow error, by default it is not called but the dataflow error is "propagated" - so it is just returned as-is. So - if I define an extension method on `Any`, it will resolve on any value, but it will return dataflow errors unchanged if called on it. In fact you can call `(Error.throw "X").sgdjkkgjkdgjkdfkjgfjk` and this will be a valid call and will return the `Error "X"` - that is because if we are holding an error, there's no way to tell what "would be" instead of it if there wasn't an error and what methods could that have - so all methods are allowed and they just act as identity.

But sometimes, I don't want the call to just fall through. To avoid this, I can explicitly add an extension method on the `Error` type. Then such methods will take precedence on error instances and allow me to customize the behaviour.

So in our Test suite, it is customary to define every extension method both on `Any` and on `Error`. That way if a value returns an `Error`, we can get an information that a test failed. It is very important because if you do not define it on `Error`, then a call `(Error.throw "X").should_equal "Y"` would fall-through and return the dataflow error - but if it's result is nowhere used - that will just be silently ignored - so a test will pass even if it's wrong - that's very bad for a test suite.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso [hunk 2]
```diff
@@ -284,17 +284,6 @@ type Time_Of_Day
...
-    ## Format this time of day as text using the default formatter.
-
-       > Example
-         Convert the current time to text.
-
-             from Standard.Base import Time_Of_Day
-
-             example_to_text = Time_Of_Day.now.to_text
-    to_text : Text
-    to_text self = @Builtin_Method "Time_Of_Day.to_text"
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso:296]: Other date/time types do not have this.

Having it breaks the `to_text` invocation on type.

I.e. `Time_Of_Day.to_text` now works. With that shadow definition it wasn't working properly.

It's a bit of a workaround to #1538

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Dialect.enso [hunk 3]
```diff
@@ -4,7 +4,9 @@ import Standard.Base.Errors.Illegal_State.Illegal_State
 import Standard.Base.Errors.Unimplemented.Unimplemented
 
 import Standard.Table.Data.Aggregate_Column.Aggregate_Column
+import Standard.Table.Data.Type.Value_Type.Value_Type
 import Standard.Table.Internal.Naming_Helpers.Naming_Helpers
+import Standard.Table.Internal.Java_Exports
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Dialect.enso]: are these changes needed? didn't seem them in the code.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso [hunk 4]
```diff
@@ -17,7 +17,7 @@ import project.Internal.Widget_Helpers
 from project.Data.Table import print_table
 from project.Data.Type.Value_Type import Value_Type, Auto
 from project.Data.Type.Value_Type_Helpers import ensure_valid_parse_target
-from project.Errors import No_Index_Set_Error, Floating_Point_Equality
+from project.Errors import No_Index_Set_Error, Floating_Point_Equality, Invalid_Value_Type
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso:20]: not used but assume should be ` ! Invalid_Value_Type` on the functions?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso [hunk 5]
```diff
@@ -908,6 +908,34 @@ type Column
...
+
+    ## Gets the day of the month as a number (1-31) from the date stored in the
+       column.
+
+       Applies only to columns that hold the `Date` or `Date_Time` types.
+       Returns a column of `Integer` type.
+    day : Column -> Column
+    day self = Value_Type.expect_has_date self.value_type related_column=self.name <|
+        new_name = Naming_Helpers.function_name "day" [self]
+        run_vectorized_unary_op self "day" (_ -> Error.throw (Illegal_State.Error "Missing vectorized implementation for `day`. This is a bug in the Table library.")) new_name
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: I guess inevitable but feels like these 3 could be extracted to a private function here.

**Review by jdunkerley — APPROVED** [review]
Couple of nits. But LGTM

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Dialect.enso (hunk 3 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Dialect.enso]: You are right.
They were used by the custom Statement_Setter implementation, I then moved it to be the default one and forgot to clean them.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso (hunk 5 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: ok

**Review by 4e6 — APPROVED** [review]
